aips/aip-x.md

```diff
@@ -0,0 +1,707 @@
+    httpOnly: true,
+    sameSite: "lax",
+  });
+
+  return c.json({ data: input });
+});
+```
+
+2. On the frontend, the application should call the fields endpoint and send a request to the wallet to _Sign In_.
+
+```tsx
+export default function SignInPage() {
+  const { signIn } = useWallet();
+
+  const handleSignIn = async () => {
+    const input = await fetch("/auth/siwa");
+
+    const response: AptosSignInOutput = await signIn(input);
+
+    // ...
+  };
+
+  return (
+    <div>
+      <button onClick={handleSignIn}>Sign In</button>
+    </div>
+  );
+}
+```
+
+3. The application should create an endpoint to verify the `AptosSignInInput` constructed at the beginning of the flow to the response from the wallet. The verification typically requires the following from the response: signature, signature type, plain text message, and account public key. We use the `deserializeSignInPublicKey`, `deserializeSignInSignature`, and `verifySignInMessage` utilities from `@aptos-labs/siwa` to handle the verification logic.
+
+```ts
+import {
+  deserializeSignInPublicKey,
+  deserializeSignInSignature,
+  verifySignInMessage,
+} from "@aptos-labs/siwa";
+
+const app = new Hono();
+
+app.post("/auth/siwa/verify", (c) => {
+  const { type, signature, message, publicKey } = c.req.json();
+
+  const input = JSON.parse(getCookie(c, "siwa-input"));
+
+  const verification = verifySignInMessage(input, {
+    publicKey: deserializeSignInPublicKey(type, publicKey),
+    signature: deserializeSignInSignature(type, signature),
+    message,
+  });
+
+  if (!verification.valid) return c.json({ errors: verification.errors }, 400);
+
+  // ... create and store a session for the user
+
+  return c.json({ data: verification });
+});
+```
+
+4. Once the verification endpoint has been created and the wallet successfully responds to the application, verify the response with the new verification endpoint.
+
+```tsx
+export default function SignInPage() {
+  const { signIn } = useWallet();
+
+  const handleSignIn = async () => {
+    const input = await fetch("/auth/siwa");
+
+    const response: AptosSignInOutput = await signIn(input);
+
+    const verification = await fetch("/auth/siwa/verify", {
+      method: "POST",
+      body: JSON.stringify({
+        type: response.message.type,
+        signature: response.signature.bcsToHex().toString(),
+        message: response.message,
+        publicKey: response.publicKey.bcsToHex().toString(),
+      }),
+    });
+
+    if (!verification.ok) throw new Error("Failed to verify message");
+
+    // ... the user is signed in, redirect to application
+  };
+
+  // ...
+}
+```
+
+## Testing
+
+The new `@aptos-labs/siwa` package will have its utilities tested to ensure that validation is done correctly. Coverage is most important here. In addition, signature verification methods will be used from the `@aptos-labs/ts-sdk`, it's important that public key verification methods are supported and tested.
```

Comment:
> it's important that public key verification methods are supported and tested

What are you referring to here?

Reply:
After convo: You mean "signature verification" (which takes a message `m`, a public key `pk` and a signature `\sigma` as input and outputs either 0 or 1)